File: javaBootcamp/src/main/java/model/Dev.java
package model;

import java.util.LinkedHashSet;
import model.content.Content;
import java.util.Optional;

public class Dev {
    private final String name;
    private final LinkedHashSet<Content> subscribedContent;
    private final LinkedHashSet<Content> completedContent;

    public Dev(String name) {
        this.name = name;
        subscribedContent = new LinkedHashSet<>();
        completedContent = new LinkedHashSet<>();
    }

    public Dev(String updateName, Dev contentToCopy) {
        this.name = updateName;
        subscribedContent = contentToCopy.subscribedContent;
        completedContent = contentToCopy.completedContent;
    }

    public double getXp() {
        return completedContent.stream().mapToDouble(Content::xpAwarded).sum();
    }

    public void subscribeContent(Content content) {
        if(!completedContent.contains(content)) {
            subscribedContent.add(content);
        }
    }

    public void unsubscribeContent(Content content) {
        subscribedContent.remove(content);
    }

    public Optional<Content> nextContent() {
        return subscribedContent.stream().findFirst();
    }

    public boolean completeContent(Content content) {
        if(subscribedContent.contains(content)) {
            subscribedContent.remove(content);
            completedContent.add(content);
            return true;
        } else {
            return false;
        }
    }

    @Override
    public String toString() {
        return "(Dev :Name " + name + " :Xp " + getXp() + ')';
    }
}

File: kotlinBootcamp/src/main/kotlin/model/Dev.kt
package model

import model.content.Content
import java.util.*


class Dev {
    private val name: String
    private val subscribedContent: LinkedHashSet<Content>
    private val completedContent: LinkedHashSet<Content>

    constructor(name: String) {
        this.name = name
        subscribedContent = LinkedHashSet()
        completedContent = LinkedHashSet()
    }

    constructor(updateName: String, contentToCopy: Dev) {
        name = updateName
        subscribedContent = contentToCopy.subscribedContent
        completedContent = contentToCopy.completedContent
    }

    val xp: Double get() = completedContent.sumOf { obj: Content -> obj.xpAwarded() }

    fun subscribeContent(content: Content) {
        if (!completedContent.contains(content)) {
            subscribedContent.add(content)
        }
    }

    fun unsubscribeContent(content: Content) {
        subscribedContent.remove(content)
    }

    fun nextContent(): Content? {
        return subscribedContent.firstOrNull()
    }

    fun completeContent(content: Content): Boolean {
        return if (subscribedContent.contains(content)) {
            subscribedContent.remove(content)
            completedContent.add(content)
            true
        } else {
            false
        }
    }

    override fun toString(): String {
        return "(Dev :Name $name :Xp $xp)"
    }
}